test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py

```diff
@@ -0,0 +1,90 @@
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+params = "cpu_offload,backward_prefetch,sharding_strategy,mixed_precision"
+cpu_offload_config = [CPUOffload(offload_params=True), CPUOffload(offload_params=False)]
+backward_prefetch_config = [BackwardPrefetch.BACKWARD_PRE, BackwardPrefetch.BACKWARD_POST, None]
+sharding_strategy_config = [ShardingStrategy.SHARD_GRAD_OP, None]
+mixed_precision = ["enable_mixed_precision", None]
+
+configs = list(itertools.product(cpu_offload_config,
+                                 backward_prefetch_config,
+                                 sharding_strategy_config,
+                                 mixed_precision))
+test_name_mapping = {
+    str(CPUOffload(offload_params=True)): "offload_true",
+    str(CPUOffload(offload_params=False)): "offload_false",
+    str(BackwardPrefetch.BACKWARD_PRE): "prefetch_pre",
+    str(BackwardPrefetch.BACKWARD_POST): "prefetch_post",
+    str(ShardingStrategy.SHARD_GRAD_OP): "shard_grad_op",
+    "enable_mixed_precision": "mixed_precision"
+}
+
+subtest_name = functools.partial(subtest_name, test_name_mapping)
+
+# TODO:
+# Check if gradients are scaled by scaling factor
+# Check if overflow error is thrown
+# Check if inf skips optimizer step
+
+class TestShardedGradScalerParityWithDDP(FSDPTest):
+    def _get_init_modes_for_test(self, cpu_offload):
+        modes = [
+            FSDPInitMode.CUDA_AFTER,
+            FSDPInitMode.CUDA_BEFORE
+        ]
+        # Note that FSDPInitMode.CUDA_NEVER works currently only with CPU
+        # offload as we explicitly bring the param back to CUDA device. In
+        # general, it will not work since we try to all_gather p.data which is
+        # on CPU but NCCL only supports GPU.
+        if cpu_offload.offload_params:
+            modes.append(FSDPInitMode.CUDA_NEVER)
+
+        return modes
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(params, configs, subtest_name)
+    def test_scaler_enabled(self, cpu_offload, backward_prefetch, sharding_strategy, mixed_precision):
+        init_modes = self._get_init_modes_for_test(cpu_offload)
+        mp = MixedPrecision(
+            param_dtype=torch.float16,
+            reduce_dtype=torch.float16,
+            buffer_dtype=torch.float16,
+        ) if mixed_precision else None
+        for fsdp_init_mode in init_modes:
+            self._test_identical_outputs(
+                NestedWrappedModule,
+                fsdp_init_mode=fsdp_init_mode,
+                cpu_offload=cpu_offload,
+                backward_prefetch=backward_prefetch,
+                sharding_strategy=sharding_strategy,
+                mixed_precision=mp,
+                enable_sharded_grad_scaler=True,
```

Comment:
is there a reason we're testing sharded grad scaler when mixed precision is _not_ enabled?

Reply:
Yeah, I was using it as a baseline test to compare DDP vs FSDP (sharded grad scaler turned on and off). That way the test actually compares the params after the iterations. This was pretty useful the first few times I ran it to identify edge cases where there were bugs in the sharded grad scaler code.

Reply:
Sounds good!